src/panel/context/Explorer/ast/index.ts

```diff
@@ -1,177 +1,221 @@
+          : selectionNode.name.value
+      ];
+
+    const node: ParsedFieldNode = parsedNodemap[key]
+      ? {
+          ...parsedNodemap[key],
+          _owner: owner
+        }
+      : {
           _id: nanoid(),
           _owner: owner,
-          cacheOutcome: ctx.cacheOutcome,
-          key: fieldKey,
-          name: fieldName,
-          args: fieldArgs
+          cacheOutcome,
+          key,
+          name,
+          args
         };
-      } else {
-        node = map[fieldKey] = copyFieldNode(map[fieldKey], owner);
-        node.cacheOutcome = ctx.cacheOutcome;
-      }
 
-      if (
-        fieldNode.selectionSet !== undefined &&
-        typeof fieldValue === "object" &&
-        fieldValue !== null
-      ) {
-        const childValue = fieldValue as Data | Data[];
-        const fieldSelection = getSelectionSet(fieldNode);
-
-        if (Array.isArray(childValue)) {
-          const size = childValue.length;
-          node.children = Array.isArray(node.children)
-            ? node.children
-            : new Array(size);
-          node.children.length = size;
-
-          for (let i = 0; i < size; i++) {
-            const childData: Data | null = childValue[i];
-            if (childData === null) {
-              node.children[i] = null;
-            } else {
-              const childMap = node.children[i] || Object.create(null);
-              node.children[i] = copyFromData(
-                ctx,
-                childMap,
-                fieldSelection,
-                childData,
-                owner
-              );
-            }
-          }
-        } else {
-          const childMap = node.children || Object.create(null);
-          node.children = copyFromData(
-            ctx,
-            childMap,
-            fieldSelection,
-            childValue,
-            owner
-          );
+    if (selectionNode.selectionSet && Array.isArray(value)) {
+      const children = (node.children || []) as ParsedNodeMap[];
+      return {
+        ...parsedNodemap,
+        [key]: {
+          ...node,
+          children: value.map(
+            (data, i) =>
+              data === null
+                ? null
+                : parseNodes({
+                    ...copyArgs,
+                    parsedNodes: children[i],
+                    selections: selectionNode.selectionSet
+                      ? selectionNode.selectionSet.selections
+                      : [],
+                    data
+                  }),
+            {}
+          )
         }
+      };
+    }
 
-        delete node.value;
-      } else {
-        node.value = fieldValue === undefined ? null : fieldValue;
-        delete node.children;
-      }
-    } else {
-      const fragmentNode = !isInlineFragment(fieldNode)
-        ? ctx.fragments[getName(fieldNode)]
-        : fieldNode;
-      if (fragmentNode !== undefined) {
-        copyFromData(ctx, map, getSelectionSet(fragmentNode), data, owner);
-      }
+    if (
```

Comment:
Could it be helpful to pull this condition out into a var/func to help describe what's happening?